File: src/main/java/parser/RecursiveLineParser.java
package parser;

import model.Node;

import java.util.List;

public class RecursiveLineParser {

    private int position;
    private List<String> tokens;

    public Node parseLineToTree(List<String> tokens) {
        this.position = 1;
        this.tokens = tokens;
        return parse();
    }

    private Node parse() {
        Node result = new Node(currentToken());
        position++;

        while (!currentToken().equals(")")) {
            if (currentToken().equals("(")) {
                position++;
                result.addChild(parse());
            } else {
                result.addChild(new Node(currentToken()));
                position++;
            }
        }

        position++;
        return result;
    }

    private String currentToken() {
        return tokens.get(position);
    }

}

File: src/main/kotlin/parser/RecursiveLineParser.kt
package parser

import model.Node

class RecursiveLineParser {

    private var position: Int = 0
    private var tokens: List<String>? = null

    fun parseLinesToTree(tokens: List<String>): Node {
        this.position = 1
        this.tokens = tokens
        return parse()
    }

    private fun parse(): Node {
        val result = Node(currentToken())
        position++

        while (currentToken() != ")") {
            if (currentToken() == "(") {
                position++
                result.addChild(parse())
            } else {
                result.addChild(Node(currentToken()))
                position++
            }
        }

        position++
        return result
    }

    private fun currentToken(): String {
        return tokens!![position]
    }

}
